Feature: Adds Installer

adamzaninovich opened the pull request:
This adds the installer into this repo and makes the following changes 
- adds more tests to installer
- sets up coverage and credo for installer
- adds installer tests etc to circle ci config
- updates mix.exs and installer/mix.exs to work in the same repo
- changes name to app in installer options
- adds help option to installer
- improves installer docs

This should resolve  #119

adamzaninovich committed:
copies installer over from alice_new project

adamzaninovich committed:
adds formatter config and gitignore to installer/

adamzaninovich committed:
fixes error in installer/mix.exs

adamzaninovich committed:
update circleci config for installer

adamzaninovich committed:
update deps, add credo to installer

adamzaninovich committed:
add coveralls config to installer

adamzaninovich committed:
increase installer test coverage to 100

adamzaninovich committed:
use project version in installer

adamzaninovich committed:
change name options to app

adamzaninovich committed:
better installer readme

adamzaninovich committed:
new handlers are setup to work with alice.console

adamzaninovich committed:
add test for installer help option

adamzaninovich committed:
remove broken readme page in docs

adamzaninovich committed:
update release script with installer notes

NateBarnes commented on installer/lib/alice_new/utilities.ex:
```diff
@@ -0,0 +1,73 @@
+defmodule AliceNew.Utilities do
+  @moduledoc false
```
Why no moduledoc?

adamzaninovich replied:
Because this is an internal module, not meant to be part of the public API.

NateBarnes commented on installer/lib/alice_new/utilities.ex:
```diff
@@ -0,0 +1,73 @@
+      :ok
+    else
+      Mix.raise(
+        "Alice v#{alice_version()} requires at least Elixir v1.7.\n " <>
```
Should this be tied to the current version like this? This could create a situation where it says "Alice 0.5.2 requires at least Elixir v1.7", so you downgrade alice only to find the previous version gives the same message. Instead shouldn't this say "0.4.0 and greater"? Since that's when we made the change that requires 1.7?

adamzaninovich replied:
what about removing the version altogether?
```suggestion
        "Alice requires at least Elixir v1.7.\n " <>
```

NateBarnes commented on installer/lib/mix/tasks/alice_new_handler.ex:
```diff
@@ -0,0 +1,96 @@
+        Utilities.check_mod_name_validity!(module_name)
+        module = Utilities.handler_module(module_name)
+
+        unless path == "." do
+          Utilities.check_directory_existence!(path)
+          File.mkdir_p!(path)
+        end
+
+        File.cd!(path, fn ->
+          HandlerGenerator.generate(app, handler_name, module, path)
+        end)
```
Same with these lines, shouldn't they be extracted into a function that takes a path?

adamzaninovich replied:
The first one, sure, I'll move it to Utilities. For the second, I don't want the HandlerGenerator called in Utilities. I think it's fine here.

NateBarnes commented on .circleci/config.yml:
```diff
@@ -4,55 +4,111 @@
+      - run:
+          name: "Check Credo"
+          command: mix credo
+      # - run:
+      #     name: "Building Dialyzer PLT"
+      #     environment:
+      #       MIX_ENV: dev
+      #     command: mix dialyzer --plt
+      #     no_output_timeout: 30m
+      # - save_cache:
+      #     name: "Save Build & PLT cache"
+      #     key: v1-alice-build-and-plt-cache-{{ .Branch }}-{{ checksum "build-branch-sha" }}
+      #     paths:
+      #       - _build
+      #       - ~/.mix
+      # - run:
+      #     name: "Run dialyzer"
+      #     environment:
+      #       MIX_ENV: dev
+      #     command: mix dialyzer --halt-exit-status
```
Same here, should be deleted rather than commented out

NateBarnes commented on .circleci/config.yml:
```diff
@@ -4,55 +4,111 @@
-            - v1-alice-build-and-plt-cache-master
-      - run: mix do deps.get, compile  # get updated dependencies & compile them
-      - run: mix compile --force --warnings-as-errors # compile project and treat warnings as error
+      # - restore_cache:
+      #     name: "Restore Build & PLT cache"
+      #     keys:
+      #       - v1-alice-build-and-plt-cache-{{ .Branch }}-{{ checksum "build-branch-sha" }}
+      #       - v1-alice-build-and-plt-cache-{{ .Branch }}
+      #       - v1-alice-build-and-plt-cache-master-{{ checksum "latest-master-sha" }}
+      #       - v1-alice-build-and-plt-cache-master
```
This should likely be deleted rather than commented out.

adamzaninovich replied:
Yeah I want to do that before merging this, but before that, I need to make a branch this this stuff in it

NateBarnes commented on installer/lib/mix/tasks/alice_new_handler.ex:
```diff
@@ -0,0 +1,96 @@
+        app = "alice_#{handler_name}"
+        Utilities.check_handler_name!(handler_name, !opts[:app])
+
+        module_name = opts[:module] || Macro.camelize(handler_name)
+        Utilities.check_mod_name_validity!(module_name)
+        module = Utilities.handler_module(module_name)
```
For some reason I think this might be better encapsulated in a single function in Utilities. Something like:
```elixir
def generate_handler_module(module_name) do
  module_name
  |> check_mod_name_validity!
  |> handler_module
end
```
That way on this side, you're just generating variables and then passing them into a single function that does both the verification and the transformation.

adamzaninovich replied:
I think Utilities is a bad name, it's not meant to be a general assortment of functions. Maybe I should have made them all private functions in this module for now, but the idea was to put functions there that aren't specific to the handler task.

The next step for the installer is making a bot generator, so the idea was to put common functions in Utilities, but since we don't have that yet, it makes the line hard to draw. I guess this makes the module yagni.

Aside from that, I don't always think it's necessary to extract a function unless it's going to be used in more than one place or unless the code is confusing and needs a name. In this case, this might be better than extracting a function:

```suggestion
        module_name = opts[:module] || Macro.camelize(handler_name)

        module =
          module_name
          |> Utilities.check_mod_name_validity!()
          |> Utilities.handler_module()
```

NateBarnes replied:
That’s certainly better than it was before, thought DRY is usually the secondary reason for extraction IMO. It’s usually to put a meaningful name on it as the primary reason.

NateBarnes replied:
The primary purpose of extraction is abstraction. Or as Rich Hickey would put it. “To decomplect” 😄

NateBarnes commented on installer/test/mix/tasks/alice_new_handler_test.exs:
```diff
@@ -0,0 +1,131 @@
+          Handler.run([@handler_name])
+
+          assert_file("#{@app_name}/README.md", fn file ->
+            assert file =~ "AliceSuperAwesome"
+            assert file =~ ~s({:alice, "~> #{@alice_version}"})
+          end)
+
+          assert_file("#{@app_name}/.formatter.exs", fn file ->
+            assert file =~ "import_deps: [:alice]"
+          end)
+
+          assert_file("#{@app_name}/.gitignore")
+
+          assert_file("#{@app_name}/config/config.exs")
+
+          assert_file("#{@app_name}/lib/alice/handlers/#{@handler_name}.ex", fn file ->
+            assert file =~ "defmodule Alice.Handlers.#{@module_name} do"
+            assert file =~ "use Alice.Router"
+          end)
+
+          assert_file("#{@app_name}/mix.exs", fn file ->
+            assert file =~ "defmodule Alice#{@module_name}.MixProject do"
+            assert file =~ "app: :#{@app_name}"
+            assert file =~ ~s({:ex_doc, ">= 0.0.0", only: :dev, runtime: false})
+            assert file =~ ~s({:alice, "~> #{@alice_version}"})
+          end)
+
+          assert_file("#{@app_name}/test/test_helper.exs")
+
+          assert_file("#{@app_name}/test/alice/handlers/#{@handler_name}_test.exs", fn file ->
+            assert file =~ "defmodule Alice.Handlers.#{@module_name}Test do"
+            assert file =~ "use Alice.HandlerCase, handlers: Alice.Handlers.SuperAwesome"
+            assert file =~ "send_message"
+            assert file =~ "first_reply"
+          end)
+        end)
+
+      assert output =~ "Your Alice handler was created successfully."
+    end)
+  end
```
Should this be a describe block for "new handler with defaults" where each file is a separate test? If for no other reason then to help isolate things when something does break? Something makes me instinctive shy away from a test with this many asserts.

```elixir
describe "when running a new test with defaults" do
  test "it should generate README file" do
    in_tmp("new handler with defaults", fn ->
      output =
        capture_io(fn ->
          Handler.run([@handler_name])

          assert_file("#{@app_name}/README.md", fn file ->
            assert file =~ "AliceSuperAwesome"
            assert file =~ ~s({:alice, "~> #{@alice_version}"})
          end)
        end)
      assert output =~ "Your Alice handler was created successfully."
    end)
  end
end
```
Something like that? Or is that overkill?

adamzaninovich replied:
In ruby it's not good to have more than one assertion per test because when it fails it's difficult to tell which failed. Having multiple assertions per test is pretty common in Elixir because ExUnit does a better job telling you where the failure was.

adamzaninovich replied:
Also in ExUnit, you can't nest describe blocks, so they don't tend to be used as much overall. Rather than being used for organization like is often the case in RSpec etc, in ExUnit they are mostly used to group different tests that need a common setup block.

NateBarnes replied:
Limiting the number of asserts per test isn’t really a rubyism as a result of tooling, it’s long been a kind of best practice to avoid run on tests or the like. Not to mention the normal better readability that comes from smaller functions with explicit names.

I’m definitely not a purist who would advise for one assert per test, but I do try to only test only one logical concept per test. With push and pull based on complication of setup. This struck me as one of those kinds of tests where I can see clean logical lines in groups of asserts, counterbalanced by setup costs.

Feel free to resolve the conversation if you don’t think it’s applicable. /shrug

NateBarnes replied:
I realize that about describe blocks. I was thinking those tests all do have a common setup, it’s just being done inside the test now 😊

NateBarnes reviewed (commented):
Looks awesome! Really exciting!

![Exciting](https://media.giphy.com/media/PjplWH49v1FS0/giphy.gif)

NateBarnes reviewed (approved):
BTW, those last two items are suggestions more than anything. This PR is definitely ready to go if you don't want to action on them 👍

adamzaninovich committed:
refactor mix task and move functions to Utilities

adamzaninovich commented:
@NateBarnes I refactored the mix task and moved a bunch of functions to `Utilities`. I think this is better. What do you think?

I decided we only have the handler task for now, if we need to split up `Utilities` later, we can.

Also, I'm definitely open for suggestions for renaming `Utilities`, that's a terrible name.

adamzaninovich committed:
remove inspect call

NateBarnes reviewed (approved):
I like the changes here, looks really clean.

On the Utilities class, it really is just a toolbox of utilities, so maybe it just needs to be more specific about what kind of utilities they are? `AliceNew.FileUtilities` or `AliceNew.FileCreationUtilities` or the like?

adamzaninovich commented:
@NateBarnes cool, I like `FileUtilities` better, good call. Thanks for the review!

adamzaninovich committed:
rename Utilities to FileUtilities

also moves non-file functions to AliceNew

adamzaninovich committed:
removes commented out dialyzer support code

adamzaninovich committed:
make MixHelper.tmp_path private
